File: agent/src/test/java/com/github/kornilova_l/flamegraph/javaagent/agent/InstrumentationTest.java
package com.github.kornilova_l.flamegraph.javaagent.agent;

import com.github.kornilova_l.flamegraph.configuration.MethodConfig;
import com.github.kornilova_l.flamegraph.javaagent.TestHelper;
import com.github.kornilova_l.flamegraph.javaagent.generate.Generator;
import com.github.kornilova_l.flamegraph.javaagent.generate.test_classes.*;
import org.junit.BeforeClass;
import org.junit.Test;
import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.util.TraceClassVisitor;

import java.io.*;
import java.lang.instrument.Instrumentation;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import static com.github.kornilova_l.flamegraph.javaagent.TestHelper.removePackage;


public class InstrumentationTest {
    private static AgentConfigurationManager configurationManager;
    private static List<MethodConfig> methodConfigs = new ArrayList<>();
    private static AgentConfigurationManager configurationManagerSaveParams;
    private static List<MethodConfig> methodConfigsSaveParams = new ArrayList<>();
    private static AgentConfigurationManager configurationManagerSaveReturn;
    private static List<MethodConfig> methodConfigsSaveReturn = new ArrayList<>();

    @BeforeClass
    public static void setup() {
        configurationManager = createConfig("*.*(*)", methodConfigs);
        configurationManagerSaveParams = createConfig("*.*(*+)", methodConfigsSaveParams);
        configurationManagerSaveReturn = createConfig("*.*(*)+", methodConfigsSaveReturn);
    }

    private static AgentConfigurationManager createConfig(String config,
                                                          List<MethodConfig> methodConfigs) {
        TestHelper.createDir("actual");
        List<String> methodConfigsStrings = new LinkedList<>();
        methodConfigsStrings.add(config);
        methodConfigsStrings.add("!*.<init>(*)");
        AgentConfigurationManager configurationManager = new AgentConfigurationManager(
                methodConfigsStrings
        );
        methodConfigs.addAll(configurationManager.findIncludingConfigs(
                "com/github/kornilova_l/flamegraph/javaagent/generate/test_classes/OneMethod"));
        return configurationManager;
    }

    @Test
    public void basicInstrumentation() {
        /* All test fail because it is not possible to generate exactly same bytecode */
//        classTest(OneMethod.class, configurationManager, methodConfigs, true);
        classTest(ThrowsException.class, ThrowsExceptionExpected.class, configurationManager, methodConfigs, true);
//        classTest(UsesThreadPool.class, configurationManager, methodConfigs);
//        classTest(SeveralReturns.class, configurationManager, methodConfigs, true);
//        classTest(TwoMethods.class, configurationManager, methodConfigs, true);
    }

    // TODO: check other tests

    @Test
    public void saveParameters() {
        /* All test fail because it is not possible to generate exactly same bytecode */
        classTest(SaveParameters.class, SaveParameters.class, configurationManagerSaveParams, methodConfigsSaveParams, true);
//        classTest(SaveSecondParam.class, configurationManagerSaveSecondParam, methodConfigsSaveSecondParam, true);
    }

    @Test
    public void saveReturnValue() {
        /* All test fail because it is not possible to generate exactly same bytecode */
        classTest(SaveReturnValue.class, SaveReturnValueExpected.class, configurationManagerSaveReturn, methodConfigsSaveReturn, true);
    }

    @Test
    public void useProxy() {
        /* All test fail because it is not possible to generate exactly same bytecode */
        classTest(UseProxy.class, UseProxyExpected.class, configurationManagerSaveReturn, methodConfigsSaveReturn, false);
    }

    @Test
    public void hasCatch() {
        classTest(HasCatch.class, HasCatchExpected.class, configurationManagerSaveReturn, methodConfigsSaveReturn, true);
    }

    private void classTest(Class testedClass,
                           Class expectedClass,
                           AgentConfigurationManager configurationManager,
                           List<MethodConfig> methodConfigs,
                           boolean hasSystemCL) {
        Generator.generate(expectedClass);
        try {
            String fullName = testedClass.getName();
            String fileName = removePackage(fullName);
            InputStream inputStream = Instrumentation.class.getResourceAsStream(
                    "/" + fullName.replace('.', '/') + ".class");
            byte[] bytes = TestHelper.getBytes(inputStream);
            ClassReader cr = new ClassReader(bytes);
            ClassWriter cw = new ClassWriter(cr, ClassWriter.COMPUTE_FRAMES);
            cr.accept(
                    new ProfilingClassVisitor(
                            cw,
                            fullName.replace('.', '/'),
                            hasSystemCL,
                            methodConfigs,
                            configurationManager
                    ), ClassReader.SKIP_FRAMES | ClassReader.SKIP_DEBUG
            );

            bytes = cw.toByteArray();

            saveClass(bytes, fileName);

            cr = new ClassReader(bytes);
            cw = new ClassWriter(cr, 0);
            File outFile = new File("src/test/resources/actual/" + fileName + ".txt");
            cr.accept(
                    new TraceClassVisitor(cw, new PrintWriter(
                            new FileOutputStream(outFile)
                    )), ClassReader.SKIP_DEBUG
            );

            TestHelper.compareFiles(new File("src/test/resources/expected/" + removePackage(expectedClass.getName()) + ".txt"),
                    outFile);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    private void saveClass(byte[] bytes, String fileName) {
        try (OutputStream outputStream = new FileOutputStream(
                new File("src/test/resources/actual/" + fileName + ".class")
        )) {
            outputStream.write(bytes);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: agent/src/test/java/com/github/kornilova_l/flamegraph/javaagent/agent/InstrumentationTest.kt
package com.github.kornilova_l.flamegraph.javaagent.agent

import com.github.kornilova_l.flamegraph.configuration.MethodConfig
import com.github.kornilova_l.flamegraph.javaagent.compareFiles
import com.github.kornilova_l.flamegraph.javaagent.createDir
import com.github.kornilova_l.flamegraph.javaagent.generate.Generator
import com.github.kornilova_l.flamegraph.javaagent.generate.test_classes.*
import com.github.kornilova_l.flamegraph.javaagent.getBytes
import com.github.kornilova_l.flamegraph.javaagent.removePackage
import org.junit.Before
import org.junit.Test
import org.objectweb.asm.ClassReader
import org.objectweb.asm.ClassWriter
import org.objectweb.asm.util.TraceClassVisitor
import java.io.File
import java.io.FileOutputStream
import java.io.PrintWriter
import java.util.*


class InstrumentationTest {
    @Before
    fun setup() {
        configurationManager = createConfig("*.*(*)", methodConfigs)
        configurationManagerSaveParams = createConfig("*.*(*+)", methodConfigsSaveParams)
        configurationManagerSaveReturn = createConfig("*.*(*)+", methodConfigsSaveReturn)
    }

    @Test
    fun methodThrowsException() {
        classTest(ThrowsException::class.java, configurationManager!!, methodConfigs)
        /* If instrumentation changed uncomment following and check it manually */
//         classTest(ThrowsException::class.java, ThrowsExceptionExpected::class.java, configurationManager, methodConfigs, true)
    }

    @Test
    fun saveParameters() {
        classTest(SaveParameters::class.java, configurationManagerSaveParams!!, methodConfigsSaveParams)
        /* If instrumentation changed uncomment following and check it manually */
//        classTest(SaveParameters::class.java, SaveParametersExpected::class.java, configurationManagerSaveParams!!, methodConfigsSaveParams, true)
    }

    @Test
    fun saveReturnValue() {
        classTest(SaveReturnValue::class.java, configurationManagerSaveReturn!!, methodConfigsSaveReturn)
        /* If instrumentation changed uncomment following and check it manually */
//        classTest(SaveReturnValue::class.java, SaveReturnValueExpected::class.java, configurationManagerSaveReturn!!, methodConfigsSaveReturn, true)
    }

    @Test
    fun useProxy() {
        classTest(UseProxy::class.java, configurationManagerSaveReturn!!, methodConfigsSaveReturn, false)
        /* If instrumentation changed uncomment following and check it manually */
//        classTest(UseProxy::class.java, UseProxyExpected::class.java, configurationManagerSaveReturn!!, methodConfigsSaveReturn, false)
    }

    @Test
    fun systemClassesTest() {
        val configurationManager = AgentConfigurationManager(listOf("${SystemClass::class.java.name}.method(*)"))
        val methodConfigs = listOf(MethodConfig(SystemClass::class.java.name, "method", "(*)"))
        classTest(SystemClass::class.java, SystemClassExpected::class.java, configurationManager, methodConfigs, false, true)
    }

    @Test
    fun fileOutputStream() {
        val methodConfigs = listOf(MethodConfig("java.io.FileOutputStream", "write", "(byte[])"))
        val configurationManager = AgentConfigurationManager(listOf("java.io.FileOutputStream.write(byte[])"))
        classTest(FileOutputStream::class.java, FileOutputStream::class.java, configurationManager, methodConfigs, false, true)
    }

    @Test
    fun hasCatch() {
        classTest(HasCatch::class.java, configurationManagerSaveReturn!!, methodConfigsSaveReturn)
        /* If instrumentation changed uncomment following and check it manually */
//        classTest(HasCatch::class.java, HasCatchExpected::class.java, configurationManagerSaveReturn!!, methodConfigsSaveReturn)
    }

    @Test
    fun hasIf() {
        classTest(HasIf::class.java, configurationManager!!, methodConfigs)
        /* If instrumentation changed uncomment following and check it manually */
//        classTest(HasIf::class.java, HasIfExpected::class.java, configurationManager!!, methodConfigs)
    }

    private fun classTest(testedClass: Class<*>,
                          expectedClass: Class<*>,
                          configurationManager: AgentConfigurationManager,
                          methodConfigs: List<MethodConfig>,
                          hasSystemCL: Boolean = true,
                          isSystemClass: Boolean = false) {
        val expectedBytecodeFile = Generator.generate(expectedClass)
        instrumentAndCompare(testedClass, configurationManager, methodConfigs, expectedBytecodeFile, hasSystemCL,
                isSystemClass, true)
    }

    private fun instrumentAndCompare(testedClass: Class<*>,
                                     configurationManager: AgentConfigurationManager,
                                     methodConfigs: List<MethodConfig>,
                                     expectedBytecodeFile: File,
                                     hasSystemCL: Boolean,
                                     isSystemClass: Boolean,
                                     deleteExpectedFile: Boolean = false) {
        val bytes = instrumentClass(testedClass, configurationManager, methodConfigs, hasSystemCL, isSystemClass)
        val fileName = removePackage(testedClass.name)

        val outFile = File("src/test/resources/actual/$fileName.txt")
        printInstructionsToFile(bytes, outFile)

        compareFiles(expectedBytecodeFile, outFile, deleteExpectedFile)
    }

    private fun classTest(testedClass: Class<*>,
                          configurationManager: AgentConfigurationManager,
                          methodConfigs: List<MethodConfig>,
                          hasSystemCL: Boolean = true,
                          isSystemClass: Boolean = false) {
        val expectedBytecodeFile = File("src/test/resources/expected/" + removePackage(testedClass.name) + ".txt")
        instrumentAndCompare(testedClass, configurationManager, methodConfigs, expectedBytecodeFile, hasSystemCL,
                isSystemClass)
    }

    private fun printInstructionsToFile(bytes: ByteArray, outFile: File) {
        val cr = ClassReader(bytes)
        val cw = ClassWriter(cr, 0)
        cr.accept(
                TraceClassVisitor(cw, PrintWriter(
                        FileOutputStream(outFile)
                )), ClassReader.SKIP_DEBUG
        )
    }

    private fun instrumentClass(testedClass: Class<*>,
                                configurationManager: AgentConfigurationManager?,
                                methodConfigs: List<MethodConfig>,
                                hasSystemCL: Boolean,
                                isSystemClass: Boolean = false): ByteArray {
        val fullName = testedClass.name

        val bytes = getBytes(testedClass)
        val cr = ClassReader(bytes)
        val cw = ClassWriter(cr, ClassWriter.COMPUTE_FRAMES)
        val traceClassVisitor = TraceClassVisitor(
                cw,
                PrintWriter(System.out)
        )
        cr.accept(
                ProfilingClassVisitor(
                        traceClassVisitor,
                        fullName.replace('.', '/'),
                        hasSystemCL,
                        methodConfigs,
                        configurationManager,
                        isSystemClass
                ),
                ClassReader.SKIP_FRAMES or ClassReader.SKIP_DEBUG
        )

        return cw.toByteArray()
    }

    companion object {
        private var configurationManager: AgentConfigurationManager? = null
        private val methodConfigs = ArrayList<MethodConfig>()
        private var configurationManagerSaveParams: AgentConfigurationManager? = null
        private val methodConfigsSaveParams = ArrayList<MethodConfig>()
        private var configurationManagerSaveReturn: AgentConfigurationManager? = null
        private val methodConfigsSaveReturn = ArrayList<MethodConfig>()

        private fun createConfig(config: String,
                                 methodConfigs: MutableList<MethodConfig>): AgentConfigurationManager {
            createDir("actual")
            val methodConfigsStrings = LinkedList<String>()
            methodConfigsStrings.add(config)
            methodConfigsStrings.add("!*.<init>(*)")
            val configurationManager = AgentConfigurationManager(
                    methodConfigsStrings
            )
            methodConfigs.addAll(configurationManager.findIncludingConfigs(
                    "com/github/kornilova_l/flamegraph/javaagent/generate/test_classes/OneMethod", false))
            return configurationManager
        }
    }
}
